ucb-rit/coldfront
# Support requests to re-send external identity linking emails

Fixes #297

Includes contributions by @jofeinstein and myself.

**Changes**
- Added a section to the User Profile where a user can generate requests for external identity linking emails to be re-sent (to the email address stored in their `User` object).
    - The section displays the latest pending request if it exists, or the latest complete request if it exists, or nothing.
- Added API endpoints to be used by a script to fetch pending requests, send emails for them, and then mark those requests as complete:
    - `/api/identity_linking_requests`
        - `GET` supports a filter `status`, with two potential values: `Pending` and `Complete`.
        - `PATCH` allows updates to the `completion_time` and `status` fields.
        - Permissions
            - `GET` is only available to superusers and staff.
            - `PATCH` is only available to superusers.
    - `/api/users/`
        - This is used to retrieve the email address for a particular user.
- Added database models to support this workflow and a management command to create them.
- Added tests.

**How to Test**
- Review the code changes and add comments as needed.
- Ensure that the following test suites pass.
    ```
    # Install the new Pip requirement first.
    python manage.py test coldfront.api.user
    python manage.py test coldfront.core.user
    ```
- Manual Testing
    - Re-run the Ansible playbook, or just the following commands:
        ```
        pip install -r requirements.txt
        python manage.py migrate
        python manage.py add_default_user_choices
        ```
    - Create some pending requests in various users' User Profiles.
    - Inspect the created requests either in the database or in the REST API.
        - To access the REST API, create a token for a superuser:
            ```
            from coldfront.core.user.models import ExpiringToken
            user = ... # Ensure this user has is_superuser set to True.
            ExpiringToken.objects.create(user=user)
            ```
        - Pass the token in the request headers: `Authorization: Token YOUR_TOKEN` (e.g., via cURL or Postman).
    - Set the `completion_time` and `status`, and ensure that the section in the User Profile is updated.

**Notes**
- On production, in `main.yml` for Ansible, `ip_range_with_api_access` will need to be updated to allow access to the relevant server.
